File: 08_19_2024_2_keys_keyboard.py
class Solution:
    def solve(self, n, s, c):
        if s == n:
            return 0
        if s > n:
            return float('inf')

        paste = 1 + self.solve(n, s + c, c)

        copy = float('inf')
        if c < s:
            copy = 1 + self.solve(n, s, s)

        return min(paste, copy)

    def minSteps(self, n):
        if n == 1:
            return 0
        return 1 + self.solve(n, 1, 1)

File: 08_19_2024_2_keys_keyboard.cpp
class Solution {
public:

    int solve(int n, int s , int c){
        if(s==n) return 0;
        if(s>n) return 1e9;

        int paste = 1 + solve(n,s+c,c);

        int copy = 1e9 ;
        if(c<s) copy = 1 + solve(n,s,s);

        return min(paste,copy);
    }

    int minSteps(int n) {
       if(n==1) return 0;
       return 1+solve(n,1,1);
    }
};
